lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,293 @@
+func (d *dir) HasVolume(volType VolumeType, volName string) bool {
+	if shared.PathExists(GetVolumeMountPoint(d.name, volType, volName)) {
+		return true
+	}
+
+	return false
+}
+
+func (d *dir) Create() error {
+	// WARNING: The Create() function cannot rely on any of the struct attributes being set.
+
+	// Set default source if missing.
+	if d.config["source"] == "" {
+		d.config["source"] = GetPoolMountPoint(d.name)
+	}
+
+	if !shared.PathExists(d.config["source"]) {
+		return fmt.Errorf("Source path '%s' doesn't exist", d.config["source"])
+	}
+
+	// Check that the path is currently empty.
+	isEmpty, err := shared.PathIsEmpty(d.config["source"])
+	if err != nil {
+		return err
+	}
+
+	if !isEmpty {
+		return fmt.Errorf("Source path '%s' isn't empty", d.config["source"])
+	}
+
+	return nil
+}
+
+func (d *dir) Delete(op *operations.Operation) error {
+	// On delete, wipe everything in the directory.
+	err := wipeDirectory(GetPoolMountPoint(d.name))
+	if err != nil {
+		return err
+	}
+
+	// Unmount the path.
+	_, err = d.Unmount()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *dir) Mount() (bool, error) {
+	path := GetPoolMountPoint(d.name)
+
+	// Check if we're dealing with an external mount.
+	if d.config["source"] == path {
+		return false, nil
+	}
+
+	// Check if already mounted.
+	if sameMount(d.config["source"], path) {
+		return false, nil
+	}
+
+	// Setup the bind-mount.
+	err := tryMount(d.config["source"], path, "none", unix.MS_BIND, "")
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (d *dir) Unmount() (bool, error) {
+	path := GetPoolMountPoint(d.name)
+
+	// Check if we're dealing with an external mount.
+	if d.config["source"] == path {
+		return false, nil
+	}
+
+	// Unmount until nothing is left mounted.
+	return forceUnmount(path)
+}
+
+func (d *dir) GetResources() (*api.ResourcesStoragePool, error) {
+	// Use the generic VFS resources.
+	return vfsResources(GetPoolMountPoint(d.name))
+}
+
+// DeleteVolume deletes a volume of the storage device. If the volume supplied is a snapshot and
+// it is the last one being removed for the parent volume then the snapshots directory for that
+// parent volume is removed too.
+func (d *dir) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	volPath := GetVolumeMountPoint(d.name, volType, volName)
+
+	if !shared.PathExists(volPath) {
+		return nil
+	}
+
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		return err
+	}
+
+	err = d.deleteQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	err = os.RemoveAll(volPath)
+	if err != nil {
+		return err
+	}
+
+	// If the volume being removed is a snapshot, check if it is the last one being removed for
+	// the volume parent and if so then remove the parent volume's snaphots directory.
+	if shared.IsSnapshot(volName) {
+		// Extract just the snapshot name from the volume name and then remove that suffix
```

Comment:
Looks to me like this entire logic isn't `dir` specific, it's effectively just about removing a directory if it's now empty, should be its own helper function so other drivers can use it.

Reply:
@stgraber fair point, will do.